### ISSUE TITLE
[API Proposal]: Add CallerType attribute to tell the method the current type that is invoking the current member.

### ISSUE THREAD

**Issue by nathan130200** [dotnet/runtime#102725]
### Background and motivation

As the title says, the introduction of the attribute `CallerTypeAttribute` which will provide the method with the the type that is invoking the current member, as well as `CallerMemberNameName`.

I had this problem while writing a micro logger to handle tasks with more logging details. And when I realized that I was either creating a gigantic work-around or creating scoped loggers that store the methodbase instance or method name + type name. 

And as .NET does not support expansive macros syntax like C++, it is not possible to simply write a macro that calls all the necessary functions by passing `MethodBase.GetCurrentMethod()` in each of them. (it would be something like `#define INVOKE_LOG(LogInst, Msg) LogInst.WriteLine(Msg, MethodBase.GetCurrentMethod())`)

### API Proposal

```csharp
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
    public sealed class CallerTypeAttribute : Attribute
    {
        public CallerTypeAttribute();
    }
}
```

### API Usage

```cs
static class Utilities
{

    void WriteToStdout([CallerType] Type currentType = default, [CallerMemberName] string memberName = default)
    {
        Console.WriteLine("[{0}::{1}]: invoked!", currentType.Name, memberName);
    }

}
```

The value of `currentType` will be filled with `typeof(MyType)` (or `typeof(Utilities)` in API usage example).

**Comment by jkotas** [comment]
Thank you for filling this suggestion. This is request for a new C# language feature. New language features are tracked in https://github.com/dotnet/csharplang repo. The attribute to support the C# language feature is secondary. We add such attributes to dotnet/runtime only once the language feature design is finalized.

This is duplicate of https://github.com/dotnet/csharplang/issues/4984 that proposes a more generalized version of this proposal. Feel free to comment with your specific scenario on that proposal.

cc @jaredpar
